### ISSUE TITLE
Add Support for FreeBSD?

### ISSUE THREAD

**Issue by NeoChen1024** [intel/haxm#106]

**Comment by raphaelning** [comment]
@JayFoxRox Thanks for pitching in. @NeoChen1024 Please clarify your feature request, and perhaps explain why you think it's useful. My guess is that you want to run FreeBSD on HAXM as a guest OS. We haven't tried that yet. If you have, what error did you run into?

> As a host: Last I checked, FreeBSD had a KVM kernel module.

Thanks for the information. FreeBSD also has its own hypervisor _bhyve_, which is like KVM + QEMU combined. So before we consider porting HAXM to FreeBSD hosts, we have to ask:

1. What are the main advantages of using QEMU over bhyve?
1. Is it possible to implement the QEMU accelerator interface using bhyve API?

**Comment by NeoChen1024** [comment]
bhyve on FreeBSD lacks emulation for a normal Graphic card, and sound hardwares like Adlib, GUS. But QEMU have all of above

**Comment by AlexAltea** [comment]
It's not entirely related to the discussion:

> I assume that HAXM will eventually fade into obscurity unless it can find its niche

I believe there's such a niche: There's no cross-platform permissively-licensed hypervisor. This might sound as quite a weak argument, but reminds me of how Clang began at the time when GCC was a more mature compiler. Sure, it offered minor performance/flexibility perks (similarly to minor compatibility/flexibilty perks offered by HAXM). However, I find one of the main reasons for Clang's success to be its license, which attracted companies like Apple, Sony, and recently even Microsoft (which already maintains a compiler).

Of course, comparing compilers and hypervisors might be a big stretch: there's too little people directly involved with the latter, among other reasons. But it can teach us some lessons:
- For instance, thinking long term, the biggest obstacle to adoption by developers, is annoying them with ["yet another standard"](https://xkcd.com/927/). Honestly, the [KVM API](https://www.kernel.org/doc/Documentation/virtual/kvm/api.txt) is the most convenient/complete one, unsurprising considering how long it's been around. Adopting it would be a huge push to HAXM, since it reduces the overhead for developers. Note how this matches exactly what Clang did with GCC, as they are frequently interchangeable in makefiles with no side effects. (Anyway, this is a discussion for another day).


Back to the original issue: Although I'm not interesting in adding FreeBSD support myself, I hope the changes/refactoring coming in #108 and related patches motivate people at joining and improving HAXM.

**Comment by tuxillo** [comment]
I don't know if it already exists, but maybe a **_"Porting"_** document would be good to have? Even if it's too basic, it can be improved over time :-)

**Comment by krytarowski** [comment]
I started to work on NetBSD port last year, but I was swamped in memory management API porting. Now since the Linux version is available, I will rebase my patches and try to push it forwards.

For start, I've pushed my generic patches that I kept locally.

**Comment by AlexAltea** [comment]
> I don't know if it already exists, but maybe a "Porting" document would be good to have?

@tuxillo What kind of contents would that document have? If you mean, guidelines on how to port HAXM to another platform maybe the best location would be comments on top of the corresponding functions, e.g.:

https://github.com/intel/haxm/blob/master/include/hax_host_mem.h#L39-L51

Right now, there's several functions that are not properly documented, so porting them requires reading the implementations for other platforms which is not really ideal.

---

@tuxillo @krytarowski Probably you guys are already aware, but you two are working towards the same goal. Just pointing that fact so that you don't "duplicate/overlap work". :-)

**Comment by krytarowski** [comment]
@AlexAltea tuxillo informed me about his DragonFlyBSD porting off-GH and news about Linux port.

I had an issue that I couldn't express all the needed operations with UVM for  HAXM, especially those ones related to physical memory, probably there is need to develop dedicated kernel interfaces. My port was barely insertable into the kernel (and immediately triggering panic) as a loadable module.

I put some notes on HAXM-4-NetBSD at our wiki, but it didn't get much traction.

http://wiki.netbsd.org/users/kamil/qemu/

With a Linux port it will be now much easier to get it to work.

**Comment by krytarowski** [comment]
https://github.com/krytarowski/haxm/compare/615f5942f17b1569d9e026130002fa354c6ab76a..26d8d685171c6a56779586776a24ed3e72122a9a my original work.. very dirty scratch. I've shared it with @tuxillo .

**Comment by tuxillo** [comment]
Hi guys,

@AlexAltea thanks for sending the linux port upstream, I think the BSDs can benefit from it in case they decide to port it.

With regards to the porting documentation I was thinking more of a short guide, to see briefly what is expected from the target OS to have. As @krytarowski pointed out the VM API for Linux is quite different from the BSDs and even the BSDs have big differences among them. I may create a draft of such document as I go with my porting efforts for DragonFly BSD but I don't promise anything :-)

Currently I'm mostly doing an exercise of learning while trying to port it, we'll see how far I can get. I'll coordinate with @krytarowski for sure to join efforts if/when needed.

**Comment by raphaelning** [comment]
> Right now, there's several functions that are not properly documented, so porting them requires reading the implementations for other platforms which is not really ideal.

Yes, sorry about that. I think part of the problem comes from the old physical memory allocation/mapping functions, i.e. those implemented by `platforms/windows/hax_mem_alloc.c` and `platforms/windows/hax_mm.c` (they don't even have a dedicated header file...). I think it's possible to get rid of them and use only the newer functions in `include/hax_host_mem.h`; these new functions were introduced with `CONFIG_HAX_EPT2`, and we tried our best to document them. Most of the old functions are used solely in `!CONFIG_HAX_EPT2` code paths, which are now obsolete.

However, the new functions also deal with host physical memory and may not be easy to implement for BSDs. @krytarowski @tuxillo Please feel free to propose a more generic abstraction.

**Comment by krytarowski** [comment]
I've learnt looking at the Linux code that indeed CONFIG_HAX_EPT2 is reducing the difficulty largely! I've missed it last year and died with my porting as we in general intend to use virtual memory everywhere unless we are accessing PCI/similar buses with dedicated APIs that are designed only for them.

**Comment by AlexAltea** [comment]
@krytarowski [`debian-9.5.0-amd64-xfce-CD-1.iso`](https://linuxtracker.org/index.php?page=torrent-details&id=757564028ea29e320e15422075ba3f9329e7f842&hit=1) should boot alright (at least the installer).

**Comment by krytarowski** [comment]
@AlexAltea could you please enable all logging messages in HAXM and share a dmsg log for `qemu-system-x86_64 -curses --enable-hax`? It will help me to spot what's broken on NetBSD, by comparison with a functional implementation.

**Comment by AlexAltea** [comment]
> @AlexAltea could you please enable all logging messages in HAXM and share a dmsg log for qemu-system-x86_64 -curses --enable-hax?

@krytarowski Here you go: [haxm-master-debian.zip](https://github.com/intel/haxm/files/2612604/haxm-master-debian.zip)

**Comment by krytarowski** [comment]
@AlexAltea Thank you! I got Debian to boot... but I had to patch fpudna in the kernel to workaround a behavior that might be a bug of HAXM.

**Comment by krytarowski** [comment]
#137 Experimental NetBSD host support has been submitted to review.

**Comment by krytarowski** [comment]
http://mail-index.netbsd.org/netbsd-users/2019/02/13/msg022207.html
